WT-8020 Convert WiredTiger Evergreen tests to use CMake

alisonfel opened the pull request:
**Note**: This is quite a large PR and its probably going to take some time to review. To make this easier, I've broken up the changes into small commits, each commit focusing on an individual aspect of our evergreen testing suite. I'd recommend you review this PR by reading the contents of each commit (particularly when its comes to the `evergreen.yml`).

---

**Goal**: The goal is to remove the `autoconf`/`libtool` & Windows `SCons` build systems for the next standalone WiredTiger release. Moving into the future, WiredTiger will only support the CMake build system. 

**Change**: As part of this initiative, it's time we transition our testing infrastructure away from `autoconf`/`libtool` and primarily use CMake. All the testing for our `develop` branch (and future releases) is intended to be done on a CMake build. This change is mainly done in two parts:
1. Converts all uses of the `autoconf` and `SCons` build system to CMake in our evergreen testing infrastructure.
2. Cleans up the evergreen yaml file, removing any lingering references to `autoconf` and `SCons`.

**Execution Strategy**: This change is going to be concurrently tested on a feature branch over an extended (~1 month) testing period. This PR can be reviewed concurrently over this period. The feature branch will be continuously re-based against `develop` to make sure its in a stable state before the final merge.

alisonfel commented on test/evergreen/compatibility_test_for_releases.sh:
```diff
@@ -273,14 +295,26 @@ older_release_branches=(mongodb-4.2 mongodb-4.0 mongodb-3.6)
+# CMake build support is reliably mature in newer release branches, whilst earlier revisions
+# primarily use Autoconf (note: some earlier branches may have CMake support, but these aren't
+# considered 'mature' versions.)
+declare -A build_sys
```
I've chosen the cut-off for CMake testing to be `develop` and any future release branches we add to this script, i.e. CMake can be considered the official build system in any releases from _this_ point in time.  
Even though `mongodb-5.0` and `mongodb-4.4` technically have a CMake build, I feel strongly its better to test these branches with `autoconf` since:
1.  The main build system used at the time of testing these release was `autoconf`(being the 'official' build system for those release).
2. CMake was not completely 'mature' in these brancheas. Several improvements & refactors have since been made (particularly in the lead up to this ticket) that aren't present in these releases.

lukech replied:
Sounds like a well-thought approach to me. Once this work is done, let's think about whether we'd like to have older branches adopt CMake-only builds. And if yes, how to handle backports.

alisonfel commented:
@keithbostic I just wanted to tag you into this PR so you can track the overall progress towards removing `autoconf`/`libtool`/`SCons` (feel free to review of course!)

etienneptl commented on test/evergreen/import_compatibility_test.sh:
```diff
@@ -21,11 +21,20 @@ build_branch()
+            config+="-DENABLE_SNAPPY=1 "
+            config+="-DWT_STANDALONE_BUILD=0 "
+            (mkdir build && cd build &&
+                $CMAKE $config ../. && make -j $(grep -c ^processor /proc/cpuinfo)) > /dev/null
```
Why don't we use `ninja` if possible?

alisonfel replied:
Generally I do encourage building with `ninja`, but I opted with `make` here to be consistent with the `autoconf` paths. `Ninja` is not available on our our build machines (i.e. rhel, macos), where `make` is generally pretty available and standard across the board. I wanted to keep the dependencies of this script simple, making it as portable as possible (as it could be run manually outside evergreen). 
Whilst it might be possible to test for `ninja` and then fallback to `make`, it feel it over complicates the script, where just using `make` keeps the script simple and consistent.

etienneptl replied:
Can we ask the BUILD team to install ninja?

alisonfel replied:
We can definitely do that, it would nice if ninja was installed on those machines. 
However I feel that's outside the scope of this script and ticket, since Make works pretty well here. I rather not make that a blocker here.

etienneptl replied:
I agree and created WT-8574. Can you add a FIXME-WT-8574 comment?

etienneptl commented on test/evergreen/compatibility_test_for_releases.sh:
```diff
@@ -44,12 +44,25 @@ build_branch()
+            config+="-DENABLE_SNAPPY=1 "
+            config+="-DWT_STANDALONE_BUILD=0 "
+            (mkdir build && cd build &&
+                $CMAKE $config ../. && make -j $(grep -c ^processor /proc/cpuinfo)) > /dev/null
```
Can we use `ninja` here?

alisonfel replied:
Same reason discussed in https://github.com/wiredtiger/wiredtiger/pull/7317#discussion_r769090697
Rather stick with `make` here.

etienneptl commented on test/evergreen/find_cmake.sh:
```diff
@@ -40,6 +40,7 @@ find_cmake ()
         # Some images have no cmake yet, or a broken cmake (see: BUILD-8570)
         echo "-- MAKE CMAKE --"
         CMAKE_INSTALL_DIR=$(readlink -f cmake-install)
+        if [ -d  cmake-3.11.0 ]; then rm -r cmake-3.11.0; fi
```
Can we make the CMake version (3.11) a variable in this script? It will make this script easier to update if required in the future.

alisonfel replied:
Done: f70bf23acb506305ccbfb9aff3779df788c0f0aa

etienneptl commented on test/evergreen/import_compatibility_test.sh:
```diff
@@ -21,11 +21,20 @@ build_branch()
+            config=""
+            config+="-DENABLE_SNAPPY=1 "
+            config+="-DWT_STANDALONE_BUILD=0 "
+            (mkdir build && cd build &&
```
Question about consistency, sometimes we create the `build` folder (like here) and sometimes `cmake_build` for CMake builds (in evergreen.yml). Should we try to create a folder with the same name when it comes to CMake builds?

alisonfel replied:
So yep,  for builds created in the `evergreen.yml`, we generally use` cmake_build`. 
This specific script is a bit special however since its not run by the evergreen compilation task. Its checks out the wiredtiger repository multiple times at different revision points, running autoconf and/or cmake builds. To keep the logic and paths consistent, I wanted to establish a consistent build directory between a repository using an autoconf build vs a cmake build. Since this script is running compatibility tests against between checkouts e.g. a 5.0 checkout attempting to use 4.4 binaries, the logic is much simpler if they use identical paths.

etienneptl commented on test/evergreen/run_format_configs.sh:
```diff
@@ -10,7 +10,7 @@ set -u
 
 # Walk into the test/format directory
-cd build_posix/test/format
+cd cmake_build/test/format
```
Should we `set -e` to exit if this command fails? Or even the first `cd` line 10?

alisonfel replied:
Done: e9e9a26faf8cb53cb02a27a864f5b894a2d1d6b2
I did it just around these commands. The test intends to catch failures in the subsequent format runs.

etienneptl commented on test/evergreen/configure_combinations.sh:
```diff
@@ -1,59 +1,84 @@
+-DENABLE_STATIC=0 -DENABLE_PYTHON=1
+-DENABLE_SNAPPY=1 -DENABLE_ZLIB=1 -DENABLE_LZ4=1
+-DHAVE_BUILTIN_EXTENSION_LZ4=1 -DHAVE_BUILTIN_EXTENSION_SNAPPY=1 -DHAVE_BUILTIN_EXTENSION_ZLIB=1
+-DHAVE_DIAGNOSTIC=1-DENABLE_PYTHON=1
```
`-DENABLE_PYTHON=1` is already line 45 and there is a missing space after `-DHAVE_DIAGNOSTIC=1`

etienneptl commented on test/evergreen/configure_combinations.sh:
```diff
@@ -1,59 +1,84 @@
+-DENABLE_SNAPPY=1 -DENABLE_ZLIB=1 -DENABLE_LZ4=1
+-DHAVE_BUILTIN_EXTENSION_LZ4=1 -DHAVE_BUILTIN_EXTENSION_SNAPPY=1 -DHAVE_BUILTIN_EXTENSION_ZLIB=1
+-DHAVE_DIAGNOSTIC=1-DENABLE_PYTHON=1
+-DENABLE_STRICT=1 -DENABLE_STATIC=1 -DENABLE_SHARED=0 -DWITH_PIC=1"
```
`-DENABLE_STATIC=1 -DENABLE_SHARED=0` are already present line 44

alisonfel replied:
Every line here is an individual configuration command. Hence the configure parameters used in line 44 is run separately to configuration parameters on this line. Hence `-DENABLE_STATIC=1 -DENABLE_SHARED=0` is necessary to have here for this particular configuration permutation.

etienneptl replied:
My bad, thanks for clarifying!

etienneptl commented on test/evergreen/configure_combinations.sh:
```diff
@@ -1,59 +1,84 @@
---enable-diagnostic --enable-python
---enable-strict --disable-shared"
+options="-DHAVE_DIAGNOSTIC=1
+-DENABLE_SHARED=0 -DENABLE_STATIC=1
```
Static should not be enabled, remove this one but keep the one line 45

alisonfel replied:
As discussed in the earlier comment: https://github.com/wiredtiger/wiredtiger/pull/7317#discussion_r768275060
This is a unique configuration permutation and run separately to whats on line 45.

etienneptl replied:
Yes, sorry! Resolving the comment.

etienneptl commented on test/evergreen/configure_combinations.sh:
```diff
@@ -1,59 +1,84 @@
+        [ "$1"  == *"clang.cmake"* ] && compiler="clang" || compiler="cc"
+        echo $compiler -o ./smoke ../examples/c/ex_smoke.c $cflags
+        $compiler -o ./smoke ../examples/c/ex_smoke.c  $cflags|| return 1
+        LD_LIBRARY_PATH="$insdir/lib:$insdir/lib64" ./smoke || return 1
```
Why do we need `$insdir/lib64` ?

alisonfel replied:
Depending on the OS, `lib64` is also a valid install path. On RHEL machines `lib64` is the chosen install path, `lib` for Multiarch Debian/Ubuntu hosts. 
When I implemented the install target in CMake, rather than coding in fixed install paths (i.e. `lib`, `bin`), I deferred to the CMake build system to determine the correct install paths based on the host it was installing for. More specifically, we use the paths determined by `[GNUInstallDir](https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html)`, which has a whole lot of custom handling based on the host its running on. Which means for RHEL targets we need to add `lib64`

etienneptl commented on test/checkpoint/recovery-test.sh:
```diff
@@ -4,23 +4,29 @@ set -x
 }
 
-if [ "$#" -ne 2 ]; then
+if [ "$#" -lt 2 ] || [ "$#" -gt 3 ]; then
```
Can we do `-ne 3`?

etienneptl commented on test/evergreen.yml:
```diff
@@ -203,41 +203,49 @@ functions:
-              --enable-diagnostic --with-builtins=lz4,snappy,zlib,zstd
+            . test/evergreen/find_cmake.sh
+            if [ -d cmake_build ]; then rm -r cmake_build; fi
+            mkdir -p cmake_build
```
The `-p` option might not be necessary since it should not exist or be removed. Happy to keep it as it's harmless.

alisonfel replied:
I think it ok to keep around, `-p` is good practice in case these paths change/evolve.

etienneptl commented on test/evergreen.yml:
```diff
@@ -413,51 +413,51 @@ functions:
       script: |
         set -o errexit
         set -o verbose
-        ${test_env_vars|} ./t ${many_db_args|} 2>&1
```
Have we been running `t` instead of `test_manydbs`?

alisonfel replied:
Autoconf produces a binary named `t`, whilst CMake produces a binary named `test_manydbs`. This change updates to the expected binary name in CMake.

etienneptl commented on test/evergreen.yml:
```diff
@@ -2435,22 +2435,24 @@ tasks:
-          configure_env_vars: CC=/opt/mongodbtoolchain/v3/bin/gcc CXX=/opt/mongodbtoolchain/v3/bin/g++ PATH=/opt/mongodbtoolchain/v3/bin:$PATH ADD_CFLAGS="-g -Werror"
-          posix_configure_flags: --enable-silent-rules --enable-strict --enable-diagnostic --disable-static
+          posix_configure_flags:
+            -DCMAKE_TOOLCHAIN_FILE=../cmake/toolchains/mongodbtoolchain_v3_gcc.cmake
```
I personally really like when you lay out the options this way, it makes them very easy to read, nice!

etienneptl commented on test/evergreen.yml:
```diff
@@ -2711,19 +2711,15 @@ tasks:
       - func: "compile wiredtiger"
         vars:
-          posix_configure_flags: --enable-strict --disable-shared --with-builtins=snappy,zlib,zstd
+          posix_configure_flags: -DENABLE_STRICT=1 -DENABLE_STATIC=1 -DENABLE_SHARED=0 -DHAVE_BUILTIN_EXTENSION_LZ4=1 -DHAVE_BUILTIN_EXTENSION_SNAPPY=1 -DHAVE_BUILTIN_EXTENSION_ZLIB=1
```
ZSTD should be enabled, not Z4

alisonfel replied:
Fixed: c58026942d53872667f6e386c46a160b64bb95cb

etienneptl commented on test/evergreen.yml:
```diff
@@ -2575,17 +2575,18 @@ tasks:
     commands:
       - func: "get project"
       - func: "compile wiredtiger"
+        # CC is set to the system default "clang" binary here as a workaround. Change it back
```
Can you add at the beginning of the comment `FIXME-WT-8482` please?

alisonfel replied:
Done: a23c9e95e1b28d4afeff3da9100e67db4cf318e0

etienneptl commented on test/evergreen.yml:
```diff
@@ -2805,14 +2806,16 @@ tasks:
-            -I/opt/mongodbtoolchain/v3/lib/gcc/ppc64le-mongodb-linux/8.2.0/include"
-          posix_configure_flags: --enable-diagnostic --with-builtins=lz4,snappy,zlib
+          posix_configure_flags:
+            -DHAVE_DIAGNOSTIC=1
```
Can we add `-DCMAKE_TOOLCHAIN_FILE=../cmake/toolchains/clang.cmake` as the first line here too? For consistency.

alisonfel replied:
f4cc43a3ecb8eddd9d439644b0a49b5f7ef8ddaa

etienneptl commented on test/evergreen.yml:
```diff
@@ -2805,14 +2806,16 @@ tasks:
-          posix_configure_flags: --enable-diagnostic --with-builtins=lz4,snappy,zlib
+          posix_configure_flags:
+            -DHAVE_DIAGNOSTIC=1
+            -DHAVE_BUILTIN_EXTENSION_LZ4=1
```
We need `-DWITH_PIC=1`

etienneptl replied:
Seems like we can copy and paste from line 2582 to 2589

alisonfel replied:
Its not really necessary, since its only building a shared target. Added it anyway in 8ac995589916edf337d5ec4c20bec8edf4a5abe3 to be consistent.

etienneptl commented on test/evergreen.yml:
```diff
@@ -2771,9 +2771,11 @@ tasks:
+          posix_configure_flags:
+            -DCMAKE_TOOLCHAIN_FILE=../cmake/toolchains/mongodbtoolchain_v3_gcc.cmake
+            -DCMAKE_C_FLAGS="-ggdb"
+            -DENABLE_PYTHON=1
```
Is this option required here?

etienneptl commented on test/evergreen.yml:
```diff
@@ -272,16 +272,26 @@ functions:
+              mkdir -p cmake_build
+              cd cmake_build
+              $CMAKE -DCMAKE_C_FLAGS="-DMIGHT_NOT_RUN -Wno-error" -DENABLE_PYTHON=1 -DENABLE_STRICT=1 -G "${cmake_generator|Ninja}" ../.
+              ninja lang/python/all
```
Question here, what does `lang/python/all` do? Is it the equivalent to `make ../../../lang/python/wiredtiger_wrap.c`?

alisonfel replied:
Its not quite equivalent, `lang/python/all` will compile the full Python API. Unfortunately the CMake build doesn't have a target at the same granularity that just generates `wiredtiger_wrap.c`. `lang/python/all` is the closest here.

etienneptl commented on test/evergreen.yml:
```diff
@@ -3781,31 +3781,29 @@ buildvariants:
+      -DCMAKE_PREFIX_PATH="$(pwd)/../TCMALLOC_LIB"
+      -DCMAKE_INSTALL_PREFIX=$(pwd)/LOCAL_INSTALL
+      -DCMAKE_C_FLAGS="-ggdb"
     smp_command: -j $(grep -c ^processor /proc/cpuinfo)
```
Do we still need this?

alisonfel replied:
Yep, we need the `smp_command`, some of the sub-tasks in this variant (and others) make use of the `smp_command` variable.

etienneptl commented on test/evergreen.yml:
```diff
@@ -125,7 +125,7 @@ functions:
         if [[ "${posix_configure_flags|}" =~ (tcmalloc|TCMALLOC) ]]; then
           git clone git@github.com:wiredtiger/automation-scripts.git
-          . automation-scripts/evergreen/find_gperftools.sh ${s3_access_key} ${s3_secret_key} ${build_variant} ${is_cmake_build|false}
+          . automation-scripts/evergreen/find_gperftools.sh ${s3_access_key} ${s3_secret_key} ${build_variant} true
```
Can we instead create the variable `is_cmake_build` and set it to true and leave `${is_cmake_build}`?
having a variable `is_cmake_build` will be easier to understand than a simple `true`.

alisonfel replied:
Done: 448fa179ea163847ea7d30467e23dce9385076f7

etienneptl commented:
I did a first pass and left a few comments/questions. Let me know if anything is unclear. I will do another pass soon to make sure I have not missed anything else.

alisonfel commented on test/evergreen/configure_combinations.sh:
```diff
@@ -1,59 +1,84 @@
-                $compiler -o ./smoke ../examples/c/ex_smoke.c  $cflags|| return 1
-                LD_LIBRARY_PATH=$insdir/lib ./smoke || return 1
+        eval $CMAKE_BIN $extra_config "$1" "$2" \
+                 -DCMAKE_INSTALL_PREFIX="$insdir" -G $GENERATOR ../. || return 1
```
I think `return 1` is the right way to go here. Given that is inside a function, if a failure happens, the caller to `BuildTest` wants to catch the return code and dump the failing configuration command. I believe `set -e` would exit the script, avoiding the error handling logic we have.

Siddhartha8899 commented on test/evergreen.yml:
```diff
@@ -2256,7 +2256,7 @@ tasks:
             set -o errexit
             set -o verbose
-            ${python_binary|python3} syscall.py --verbose --preserve
+            WT_BUILDDIR=$(pwd)/../../cmake_build LD_LIBRARY_PATH=$WT_BUILDDIR ${python_binary|python3} syscall.py --verbose --preserve
```
Can we use  `${test_env_vars|}` instead of `WT_BUILDDIR=$(pwd)/../../cmake_build LD_LIBRARY_PATH=$WT_BUILDDIR` here?

alisonfel replied:
Unfortunately this won't quite work. Sometimes the `test_env_vars` variable is setup with environment variables that will break the test. On specific example is using`LD_PRELOAD=/usr/local/lib/libeatmydata.so`. This is unfortunately impacts/breaks the `syscall-linux` runtime, failing the test. 
For this specific case, its better to be explicit about the variables the test requires, keeping the runtime environment as clean as possible for `syscall.py`

Siddhartha8899 commented on test/evergreen.yml:
```diff
@@ -370,16 +356,12 @@ functions:
           export "PATH=/cygdrive/c/Python39:/cygdrive/c/Python39/Scripts:$PATH"
           export "PYTHONPATH=$(pwd)/../lang/python/wiredtiger):$(cygpath -w $(pwd)/../lang/python)"
```
`PYTHONPATH` is incorrect, [WT-8508](https://jira.mongodb.org/browse/WT-8508) should fix this. [PR](https://github.com/wiredtiger/wiredtiger/pull/7288/files) will be merged in develop soon.

alisonfel replied:
I'll hold out to merge the PR in this branch, rather than change it here.

alisonfel replied:
This change should now be merged in.

Siddhartha8899 commented on test/evergreen.yml:
```diff
@@ -2886,7 +2833,7 @@ tasks:
             set -o errexit
```
Can we get rid of `configure_env_vars` [here](https://github.com/wiredtiger/wiredtiger/pull/7317/files#diff-42f20e21217bd7d4b14489056e1ea24ea2b87599c85413d49182a3b0b721c132R2826)?

alisonfel replied:
Good point, fixed in: 47a822a

alisonfel commented on test/evergreen.yml:
```diff
@@ -568,11 +568,11 @@ functions:
         set -o errexit
         set -o verbose
-        ./t ${checkpoint_args} 2>&1
+        ./test_checkpoint ${checkpoint_args} 2>&1
```
Done: 358168e3b8fa0a32dcd3867c0c16a06ba790643a

alisonfel commented:
**Feature Branch Update** (`wt-8020-cmake-evergreen-feature`): https://evergreen.mongodb.com/waterfall/wiredtiger-feature 
- Squashed PR feedback committed in this branch (`wt-8020-cmake-evergreen`): https://github.com/wiredtiger/wiredtiger/commit/cd231d4d765bf1dab249ed789938ede6c9e24f3a
- Squashed `develop`: https://github.com/wiredtiger/wiredtiger/commit/5f7995c27647a06a80f773cd5faa57aaa0714a0c

lukech commented on test/evergreen/import_compatibility_test.sh:
```diff
@@ -38,7 +47,7 @@ create_file()
 
-    wt_cmd="$1/wt"
+    wt_cmd="$1/build/wt"
     test_dir="$1/WT_TEST/"
```
I think we have a convention of having the `WT_TEST` directory created under the build directory. Not strong opinion just list here for consideration.

alisonfel replied:
Fair point. I've updated it to be under the build directory: 5a2c1a343d8f86b328d000170573c713aa43af94

lukech commented on test/evergreen/compatibility_test_for_releases.sh:
```diff
@@ -210,15 +227,17 @@ verify_branches()
-            echo "$1/wt verifying $2 access method $am..."
-            dir="$2/test/format/RUNDIR.$am"
-            WIREDTIGER_CONFIG="$EXT" ./wt $(bflag $1) -h "../$dir" verify table:wt
+            echo "$wt_bin verifying $2 access method $am..."
```
The `$1` (i.e. the branch) is useful information to be dumped out here.

alisonfel replied:
Yep, good catch. I've updated it in aa3746878f698945f508158081bab167302af757

lukech commented on test/evergreen/configure_combinations.sh:
```diff
@@ -1,59 +1,84 @@
 
-flags="CFLAGS=\"-Werror -Wall -Wextra -Waddress -Waggregate-return -Wbad-function-cast -Wcast-align -Wdeclaration-after-statement -Wformat-security -Wformat-nonliteral -Wformat=2 -Wmissing-declarations -Wmissing-field-initializers -Wmissing-prototypes -Wnested-externs -Wno-unused-parameter -Wpointer-arith -Wredundant-decls -Wshadow -Wundef -Wunused -Wwrite-strings -O -fno-strict-aliasing -Wuninit ...
-CC=clang CFLAGS=\"-Wall -Werror -Qunused-arguments -Wno-self-assign -Wno-parentheses-equality -Wno-array-bounds\""
+flags="-DCMAKE_TOOLCHAIN_FILE=$curdir/cmake/toolchains/gcc.cmake -DCMAKE_C_FLAGS=\"-Werror -Wall -Wextra -Waddress -Waggregate-return -Wbad-function-cast -Wno-missing-field-initializers -Wno-cast-function-type -Wcast-align -Wformat-security -Wformat-nonliteral -Wformat=2 -Wnested-externs -Wno-unused-parameter -Wpointer-arith -Wredundant-decls  -Wundef -Wunused -Wwrite-strings -O -fno-strict-alias ...
```
Some of the cflags are removed, e.g. `-Wdeclaration-after-statement`, `-Wmissing-declarations`, `-Wmissing-prototypes`, others are added, e.g. `-Wno-discarded-qualifiers -Wno-incompatible-pointer-types -Wno-int-conversion -Wno-sign-conversion -Wno-pointer-sign -Wno-unused-variable`. Any reason for the change?

alisonfel replied:
A lot of the rework here was due to a number of the diagnostic flags being incompatible with the Python API. The code generated by SWIG would conflict with multiple diagnostics listed in this script (as you identified). This became more present in CMake since the SWIG file is dynamically generated a compile time, as opposed to a pre-generated file in autoconf (which the above set of flags would've been fitted too).

Though thanks for raising this as I did take an opportunity to re-evaluate this change. The diagnostics are probably still useful for the main WiredTiger C codebase, where the Python API is just a special case. As such I think the build system should be aware of what diagnostics are incompatible with the code generated by SWIG. I've updated the branch with 465c3a7e88783828a09d991476a25924d66222c5. The commit:
- Restores the original diagnostics (as there were before this branch)
- Expands the CMakeLists for the Python API to ignore specific diagnostics when compiling SWIG generated code.

lukech commented on test/evergreen/configure_combinations.sh:
```diff
@@ -1,59 +1,84 @@
+-DENABLE_SNAPPY=1 -DENABLE_ZLIB=1 -DENABLE_LZ4=1
+-DHAVE_BUILTIN_EXTENSION_LZ4=1 -DHAVE_BUILTIN_EXTENSION_SNAPPY=1 -DHAVE_BUILTIN_EXTENSION_ZLIB=1
+-DHAVE_DIAGNOSTIC=1-DENABLE_PYTHON=1
+-DENABLE_STRICT=1 -DENABLE_STATIC=1 -DENABLE_SHARED=0 -DWITH_PIC=1"
```
Can you remind me why `-DWITH_PIC=1` is necessary here?

alisonfel replied:
This is needed to successfully compile the `smoke` binary in  the `BuildTest` function. The `smoke` program is built as a dynamic executable, and given we we are using a static wiredtiger library, the code needs to be build with `fPIC` (otherwise we create symbol relocation errors). I alternatively tried compiling the `smoke` binary as a static executable (when using a static wiredtiger library), but this approach didn't work on our RHEL machines since it doesn't have installed static builds of `pthread`, `dl` etc. 
I felt just enabling `WITH_PIC` was the simplest approach here.

lukech commented on test/evergreen/configure_combinations.sh:
```diff
@@ -1,59 +1,84 @@
-                echo $compiler -o ./smoke ../examples/c/ex_smoke.c $cflags
-                $compiler -o ./smoke ../examples/c/ex_smoke.c  $cflags|| return 1
-                LD_LIBRARY_PATH=$insdir/lib ./smoke || return 1
+        eval $CMAKE_BIN $extra_config "$1" "$2" \
```
Is it intentional the setting line for `$extra_config` gets removed?

alisonfel replied:
Yep, I removed `extra_config` since `--enable-silent-rules` is not necessary with CMake. Given that was its only use, I figured we didn't need it. Though I missed cleaning up the variable on this line. Fixed in 7309abee8efd0e26251f0151070d253f8a6f507c

lukech commented on test/evergreen.yml:
```diff
@@ -203,41 +203,49 @@ functions:
+            if [ -d cmake_build ]; then rm -r cmake_build; fi
+            mkdir -p cmake_build
+            cd cmake_build
+            $CMAKE -DCMAKE_TOOLCHAIN_FILE=../cmake/toolchains/mongodbtoolchain_v3_clang.cmake -DCMAKE_C_FLAGS="-ggdb" -DWITH_PIC=1 \
+              -DHAVE_DIAGNOSTIC=1 -DCMAKE_BUILD_TYPE=ASan \
+              -DHAVE_BUILTIN_EXTENSION_LZ4=1 -DHAVE_BUILTIN_EXTENSION_SNAPPY=1 -DHAVE_BUILTIN_EXTENSION_ZLIB=1 ${configure_python_setting|} \
+              -G "${cmake_generator|Ninja}" ../.
```
I'm wonder how `-fno-omit-frame-pointer` is translated into the CMake ASAN build mode. It looks the `no_omit_frame_flag` is not set for ASAN build mode in `cmake/configs/modes.cmake` file.

alisonfel replied:
Good catch, `-fno-omit-frame-pointer` should be used in the CMake ASAN build mode (I completely missed this when refactoring the build modes). I've updated it here: cec889dbbfa2009222c0bd5729199ee4af86878c

lukech commented on test/evergreen.yml:
```diff
@@ -740,7 +748,7 @@ tasks:
       - func: "compile wiredtiger"
         vars:
-          posix_configure_flags: --enable-silent-rules --enable-strict --disable-shared
+          posix_configure_flags: -DENABLE_STRICT=1 -DENABLE_SHARED=0 -DENABLE_STATIC=1
```
I'm wondering how `--enable-silent-rules` gets mapped in the CMake setting.

alisonfel replied:
Good question, CMake configured builds by default are silent (for both `ninja` and `make`), hence I didn't feel a need to specify an equivalent to `--enable-silent-rules` for our CMake. To get a verbose build, the option is deferred the actual invocation of the build system (`ninja -v` or `make VERBOSE=1`)

lukech commented on test/evergreen.yml:
```diff
@@ -349,12 +349,13 @@ functions:
-        ${test_env_vars|} ${make_command|make} VERBOSE=1 check -C ${directory}  ${smp_command|} 2>&1
+        . test/evergreen/find_cmake.sh
+        cd cmake_build
+        ${test_env_vars|} cd ${directory};$CTEST ${smp_command|} --output-on-failure 2>&1
```
Can `cd ${directory}` be moved up (maybe combine with the previous cd command, so that `${test_env_vars|}` is guaranteed to apply to the $CTEST command?

lukech replied:
Btw, will this change lose the explicit setting of `VERBOSE=1`?

alisonfel replied:
Good point, I've reworked the ordering of cd commands here: 175f33f5831dac971fd223b58e7d4d57eb0b48e8

---

Given CTest is already quite verbose, sometime ago we removed uses of the verbose option `-V` (equivalent to `VERBOSE=1`), in favor of `--output-on-failure`. This option will dump the entire log of a test if it fails, otherwise it masks the output to just summary information. My intention was to maintain that setting here.

lukech commented on test/evergreen.yml:
```diff
@@ -2165,19 +2165,18 @@ tasks:
       # Handle special build combination for running all the diagnostic tests.
       - func: "configure wiredtiger"
         vars:
-          configure_env_vars: CC=/opt/mongodbtoolchain/v3/bin/gcc CXX=/opt/mongodbtoolchain/v3/bin/g++ PATH=/opt/mongodbtoolchain/v3/bin:$PATH ADD_CFLAGS="-g -Werror"
```
I'm wondering how the removed `ADD_CFLAGS="-g -Werror"` gets translated in CMake.

alisonfel replied:
This was some small cleaning since those flags aren't really needed, `-DHAVE_DIAGNOSTIC=1` (or `--enable-diagnostic`) will set the `-g` flag, whilst `-DENABLE_STRICT=1` (`--enable-strict `) will set the `-Werror` flag.

lukech commented on test/evergreen.yml:
```diff
@@ -2165,19 +2165,18 @@ tasks:
         vars:
-          configure_env_vars: CC=/opt/mongodbtoolchain/v3/bin/gcc CXX=/opt/mongodbtoolchain/v3/bin/g++ PATH=/opt/mongodbtoolchain/v3/bin:$PATH ADD_CFLAGS="-g -Werror"
-          posix_configure_flags: --enable-silent-rules --enable-strict --enable-diagnostic --disable-static
+          posix_configure_flags: -DENABLE_STRICT=1 -DHAVE_DIAGNOSTIC=1 -DENABLE_STATIC=0
```
Is `-DENABLE_SHARED=1` required here?

alisonfel replied:
I don't believe this should be necessary, `ENABLE_SHARED` is enabled by default (hence omitting the configuration).

lukech commented on test/evergreen.yml:
```diff
@@ -272,16 +272,26 @@ functions:
+              if [ -d cmake_build ]; then rm -r cmake_build; fi
+              mkdir -p cmake_build
+              cd cmake_build
+              $CMAKE -DCMAKE_C_FLAGS="-DMIGHT_NOT_RUN -Wno-error" -DENABLE_PYTHON=1 -DENABLE_STRICT=1 -G "${cmake_generator|Ninja}" ../.
+              ninja lang/python/all
```
What if `$cmake_generator` is set to make?

alisonfel replied:
Good catch, that would be error-prone. Given that invoking the python target is different across `ninja` and `make`, to keep the logic simple, I've re-worked this task to just consistently use `make` in 5068304eb1c364179fedb5dbb3db72805aac0994 (I preferred `make` here as its generally available across our machines).

lukech commented on test/evergreen.yml:
```diff
@@ -3721,58 +3721,6 @@ tasks:
 
 buildvariants:
 
-- name: ubuntu2004
```
Should we rename the "ubuntu2004-cmake" to the new "ubuntu2004" build variant?

lukech replied:
Never mind, I saw the change in a later commit.

lukech commented on test/evergreen.yml:
```diff
@@ -4079,11 +4079,13 @@ buildvariants:
     smp_command: -j $(grep -c ^processor /proc/cpuinfo)
-    make_command: PATH=/opt/mongodbtoolchain/v3/bin:$PATH make
+    make_command: ninja
+    is_cmake_build: true
```
Just wondering do we expect most of the build variants (if not all) are going to have `is_cmake_build` set to "true" after this PR?

lukech replied:
Never mind. I found the answer in a later commit.

lukech commented on test/evergreen.yml:
```diff
@@ -4194,7 +4194,9 @@ buildvariants:
-    - name: unit-test
+    - name: ".unit_test"
+    - name: fops
+
```
Should ".windows_only" be added here?

lukech replied:
Never mind, I saw the change in a later commit.

lukech commented on test/evergreen.yml:
```diff
@@ -4169,19 +4169,6 @@ buildvariants:
-    - name: ".unit_test"
-    - name: fops
-
 - name: windows-64-cmake
```
Can we rename `windows-64-cmake` to the new `windows-64` build variant?

alisonfel replied:
Yep, sorry for the confusion. This is changed in a later commit (as you discovered in you other comments).

lukech commented on test/evergreen.yml:
```diff
@@ -4316,6 +4316,10 @@ buildvariants:
     - name: compile
     - name: make-check-test
```
The make-check-test is actually not covered by the main PPC build variant. I think stress tests are the focus of these exotic platforms (unit-tests for sanity purpose).

alisonfel replied:
Good catch, that leaked through from the original CMake variant. Fixed in: cfac8c80bf5e520c50cb3b2871ac69cae7987dc6

lukech commented on test/evergreen.yml:
```diff
@@ -4375,3 +4379,7 @@ buildvariants:
     - name: compile
     - name: make-check-test
```
Ditto

alisonfel replied:
Fixed in: cfac8c8

alisonfel commented:
I've done a full rebase of the branch to reflect the latest changes committed to `develop`.  In particular this rebase re-aligns the branch so it reflects the latest updates committed to our evergreen testing.    

To make it easier to track and avoid having to review the entire commit set again, I've listed below the commits that     
have since changed (as a consequence of rebasing on `develop`):                                                                                            
- b4936913dd4b2c265bdcfeac437dd7126920dfdc yml: Converted 'generate-datafile-x' to CMake    
- 0305af75858af77bc5423ab0a9e78ce9612f395a yml: Converted 'data-validation' tests to CMake    
- db20f3498897b57ec057adb77a6f577ab1ae919e yml: Converted misc 'format' tests to CMake    
- b292edd8b44052d43f5df2c55bc800adcd781ca8 yml: Converted 'endian' fetch artifacts to CMake    
- 0a74e72f49187d3e87b96f4716c21a5cf2b33efc yml: Removed 'ubuntu2004'    
- f8c6d584ada006f4af4a5fb0d67a54f42ec08269 yml: Converted cppsuite tests to CMake    
- 775f4d0de81c7002f85f27f4f625d9bd3c6cda53 yml: Remove autoconf/scons build logic    
    
Additionally the following commits have been added (since the last most-recent commit 8f130979b5de6ded8f89d697087901e010761a38):
- cec889dbbfa2009222c0bd5729199ee4af86878c Use -fno-omit-frame-pointer in ASAN CMake build    
- 7309abee8efd0e26251f0151070d253f8a6f507c Removed extra_config in configure_combinations.sh    
- 175f33f5831dac971fd223b58e7d4d57eb0b48e8 yml: Combine 'cd' cmd in 'make check directory'    
- aa3746878f698945f508158081bab167302af757 Log branch in compatibility_test_for_releases.sh    
- 5068304eb1c364179fedb5dbb3db72805aac0994 yml: Consistent generator in docs compile    
- 5a2c1a343d8f86b328d000170573c713aa43af94 Moved 'WT_TEST' under build dir in compat test    
- 465c3a7e88783828a09d991476a25924d66222c5 Restored diagnostics in configure_combinations.sh    
- cfac8c80bf5e520c50cb3b2871ac69cae7987dc6 yml: Remove 'make-check-test' from exotic variants    
- 46073512e4c3e5d9a705dd05e4af644d97de9973 yml: Push/Pop working dir in doc compile task

etienneptl reviewed (approved):
LGTM, this is awesome work!

alisonfel commented:
I've done a full (hopefully the last major) rebase of the branch to reflect the latest changes committed to `develop`.  In particular this rebase re-aligns the branch so it reflects the latest updates committed to our evergreen testing.    

To make it easier to track and avoid having to review the entire commit set again, I've again listed below the commits that     
have since changed (as a consequence of rebasing on `develop`):                                                                                            
- c27d520b6ab3da2098a4e78a130cc34ee021ad30 yml: Convert c/csuite tests to use CMake builds
- e3e34e0281d3bf3cd66753e33d50f713d50467c4 yml: Converted 'ftruncate-test' to use CMake
- b1554163a700f5f6c73c489e4890a1c0a8b72b6e yml: Removed 'windows-64'
- 11c42c3c96d81a8ebebe437e4ccd6413dce9dcd2 yml: Remove autoconf/scons build logic
    
Additionally the following commit has been added:
- 05377897f7f13bf36532beacfbaf82551c548bd1 yml: Removed libgcc includes for PPC format builds

lukech reviewed (approved):
LGTM. Well done!

Siddhartha8899 reviewed (approved):
LGTM!!
